test/common/router/config_impl_test.cc

```diff
@@ -1651,52 +1686,116 @@ TEST_F(RouterMatcherHashPolicyTest, HashIpv6DifferentAddresses) {
+  }
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo?hash=meh", "GET");
+    const auto hash_policy = config().route(headers, 0)->routeEntry()->hashPolicy();
+    hash_2 = hash_policy->generateHash(nullptr, headers, add_cookie_nop_).value();
+  }
+  EXPECT_NE(hash_1, hash_2);
+}
+
+TEST_F(RouterMatcherHashPolicyTest, QueryParameterHashNoSuchKey) {
+  Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo?hash=bluh&bluh=eh", "GET");
+
+  firstRouteHashPolicy()->mutable_query_parameter()->set_key("meh");
+  const auto hash_policy = config().route(headers, 0)->routeEntry()->hashPolicy();
+  EXPECT_FALSE(hash_policy->generateHash(nullptr, headers, add_cookie_nop_));
+}
+
+TEST_F(RouterMatcherHashPolicyTest, HostPathHash) {
+  firstRouteHashPolicy()->mutable_host_path()->default_instance();
+  secondRouteHashPolicy()->mutable_host_path()->default_instance();
+  uint64_t hash_1, hash_2;
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+    const auto* hash_policy = config().route(headers, 0)->routeEntry()->hashPolicy();
+    hash_1 = hash_policy->generateHash(nullptr, headers, add_cookie_nop_).value();
+  }
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/bar", "GET");
+    const auto* hash_policy = config().route(headers, 0)->routeEntry()->hashPolicy();
+    hash_2 = hash_policy->generateHash(nullptr, headers, add_cookie_nop_).value();
+  }
+
+  EXPECT_NE(hash_1, hash_2);
+}
+
 TEST_F(RouterMatcherHashPolicyTest, HashMultiple) {
   auto route = route_config_.mutable_virtual_hosts(0)->mutable_routes(0)->mutable_route();
   route->add_hash_policy()->mutable_header()->set_header_name("foo_header");
   route->add_hash_policy()->mutable_connection_properties()->set_source_ip(true);
+  route->add_hash_policy()->mutable_host_path()->default_instance();
+  route->add_hash_policy()->mutable_query_parameter()->set_key("hash");
   Network::Address::Ipv4Instance address("4.3.2.1");
 
-  uint64_t hash_h, hash_ip, hash_both;
+  std::map<uint64_t, int> hash_by_count;
   {
+    // HostPath.
     Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
     Router::RouteConstSharedPtr route = config().route(headers, 0);
-    EXPECT_FALSE(
-        route->routeEntry()->hashPolicy()->generateHash(nullptr, headers, add_cookie_nop_));
+    uint64_t h =
+        route->routeEntry()->hashPolicy()->generateHash(nullptr, headers, add_cookie_nop_).value();
+    hash_by_count[h]++;
   }
   {
+    // HostPath and Header.
     Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
     headers.addCopy("foo_header", "bar");
     Router::RouteConstSharedPtr route = config().route(headers, 0);
-    hash_h =
+    uint64_t h =
         route->routeEntry()->hashPolicy()->generateHash(nullptr, headers, add_cookie_nop_).value();
+    EXPECT_EQ(hash_by_count.end(), hash_by_count.find(h));
```

Comment:
This could just be `EXPECT_EQ(hash_by_count[h], 0)` or, if you like matchers, `EXPECT_THAT(hash_by_count, Not(Contains(h))`. Same throughout this test.

Reply:
again Sorry  this is reverted.
Meanwhile I think it's about same expressive?